lib/NeuraDialect/Transforms/CanonicalizeLiveInPass.cpp

```diff
@@ -17,24 +21,275 @@ using namespace mlir;
+    llvm::errs()
+        << "[CanoLiveIn] False branch creates a back edge (loop pattern)\n";
+    return false;
+  }
+
+  // Checks if both branches can reach using_block.
+  bool true_reaches = (true_dest == using_block);
+  if (!true_reaches) {
+    if (dom_info.dominates(true_dest, using_block)) {
+      true_reaches = true;
+    } else {
+      for (Block *pred : using_block->getPredecessors()) {
+        if (pred == true_dest || dom_info.dominates(true_dest, pred)) {
+          true_reaches = true;
+          break;
+        }
+      }
+    }
+  }
+
+  bool false_reaches = (false_dest == using_block);
+  if (!false_reaches) {
+    if (dom_info.dominates(false_dest, using_block)) {
+      false_reaches = true;
+    } else {
+      for (Block *pred : using_block->getPredecessors()) {
+        if (pred == false_dest || dom_info.dominates(false_dest, pred)) {
+          false_reaches = true;
+          break;
+        }
+      }
+    }
+  }
+
+  if (!true_reaches || !false_reaches) {
+    return false;
+  }
+
+  return true;
+}
+
+DenseMap<Block *, SmallVector<DirectDataflowLiveIn>>
+identifyDirectDataflowLiveIns(Region &region, DominanceInfo &dom_info,
+                              PostDominanceInfo &post_dom_info) {
+  DenseMap<Block *, SmallVector<DirectDataflowLiveIn>>
+      using_block_to_direct_dataflow_live_ins;
+  for (Block &block : region.getBlocks()) {
+    // Skips the entry block.
+    if (&block == &region.front()) {
+      continue;
+    }
+
+    // Collects direct live-in values for the block.
+    SetVector<Value> live_ins;
+    for (Operation &op : block.getOperations()) {
+      for (Value operand : op.getOperands()) {
+        // If the operand is defined in another block, it is a live-in value.
+        if (auto block_arg = dyn_cast<BlockArgument>(operand)) {
+          if (block_arg.getOwner() != &block) {
+            live_ins.insert(operand);
+          }
+        } else {
+          Operation *def_op = operand.getDefiningOp();
+          if (def_op && def_op->getBlock() != &block) {
+            live_ins.insert(operand);
+          }
+        }
+      }
+    }
+
+    // Checks each live-in value to see if it has direct dataflow dependency.
+    for (Value live_in : live_ins) {
+      Block *defining_block = nullptr;
+
+      if (auto block_arg = dyn_cast<BlockArgument>(live_in)) {
+        defining_block = block_arg.getOwner();
+      } else {
+        Operation *def_op = live_in.getDefiningOp();
+        if (def_op) {
+          defining_block = def_op->getBlock();
+        }
+      }
+
+      if (!defining_block) {
+        continue;
+      }
+
+      if (pathsCrossConditionalBranch(defining_block, &block, dom_info,
+                                      post_dom_info)) {
+        DirectDataflowLiveIn direct_dataflow_live_in;
+        direct_dataflow_live_in.value = live_in;
+        direct_dataflow_live_in.defining_block = defining_block;
+        direct_dataflow_live_in.using_block = &block;
```

Comment:
`defining_block` and `using_block` are not actually used? then we we include them inside the struct?